### PR TITLE
IA-2019 support deleteDisk query param for deleteRuntime API

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-2019

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Review by xuscrbw — APPROVED** [review]
Looks good. Would like someone to verify the PubsubSubscriber changes just in case

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala [hunk 1]
```diff
@@ -210,10 +217,16 @@ class RuntimeRoutes(runtimeService: RuntimeService[IO], userInfoDirectives: User
...
                                         runtimeName: RuntimeName,
+                                        params: Map[String, String],
                                         span: Span): IO[ToResponseMarshallable] = {
     val res = for {
       implicit0(ctx: ApplicativeAsk[IO, AppContext]) <- AppContext.lift[IO](Some(span))
-      _ <- runtimeService.deleteRuntime(userInfo, googleProject, runtimeName)
+      deleteDisk = params
+        .get("deleteDisk")
+        .map(s => if (s == "true") true else false)
+        .getOrElse(true) //if `deleteDisk` is explicitly set to true, then we delete disk; otherwise, we don't
```
Comment (gpcarr) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala]: why do we say `.getOrElse(true)`, if `deleteDisk` is not in `params` then wouldn't we want to set `deleteDisk` to false so we don't delete the disk?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala]: ah nice catch 👍

**Comment by Qi77Qi** [comment]
patchspec passes locally..jenkins retest

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala [hunk 2]
```diff
@@ -130,6 +139,57 @@ class HttpRoutesSpec
...
+    }
+  }
+
+  it should "not delete a runtime and disk if deleteDisk is false" in {
+    val runtimeService = new BaseMockRuntimeServiceInterp {
+      override def deleteRuntime(deleteRuntimeRequest: DeleteRuntimeRequest)(
+        implicit as: ApplicativeAsk[IO, AppContext]
+      ): IO[Unit] = IO {
+        val expectedDeleteRuntime =
+          DeleteRuntimeRequest(timedUserInfo, GoogleProject("googleProject1"), RuntimeName("runtime1"), true)
```
Comment (akarukappadath) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala]: should `deleteDisk` here be `false`?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala [hunk 3]
```diff
@@ -130,6 +139,57 @@ class HttpRoutesSpec
...
+        deleteRuntimeRequest shouldBe expectedDeleteRuntime
+      }
+    }
+    val routes = fakeRoutes(runtimeService)
+    Delete("/api/google/v1/runtimes/googleProject1/runtime1?deleteDisk=true") ~> routes.route ~> check {
```
Comment (akarukappadath) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala]: same as above - shouldn't this `deleteDisk` be `false`?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala [hunk 4]
```diff
@@ -130,6 +139,57 @@ class HttpRoutesSpec
...
+      status shouldEqual StatusCodes.Accepted
+      validateRawCookie(header("Set-Cookie"))
+    }
+  }
+
+  it should "not delete a runtime and disk if deleteDisk is false" in {
```
Comment (akarukappadath) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala]: could we rename this test to "should only delete runtime if deleteDisk is false" or "should not delete disk if deleteDisk is false"?

**Review by akarukappadath — CHANGES_REQUESTED** [review]

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala (hunk 2 above)
Comment (Qi77Qi) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala]: nice catch 👍

**Review by akarukappadath — APPROVED** [review]

**Comment by Qi77Qi** [comment]
jenkins retest

**Review by rtitle — APPROVED** [review]

**Comment by rtitle** [comment]
Looks good.. btw `gcloud` has similar functionality: you can pass a `--delete-disks` flag to `gcloud compute instances delete`.

**Comment by Qi77Qi** [comment]
@rtitle good point...I didn't find an option on `deleteRuntime` call directly, but maybe if we set disk to auto delete before making delete call, it'll work...I'll give it a try

**Comment by rtitle** [comment]
I think the current approach is fine, I didn’t see a —delete-disk option in the GCE API either. I suspect the gcloud cli does something similar where it first deletes the disks, then the instance.

**Comment by Qi77Qi** [comment]
@rtitle  setting the flag `auto-delete` to true works...see https://github.com/broadinstitute/workbench-libs/pull/323, tested in console

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala [hunk 5]
```diff
@@ -123,19 +123,21 @@ class HttpSamDAO[F[_]: Effect](httpClient: Client[F], config: HttpSamDaoConfig,
       _ <- logger.info(
         s"${traceId} | creating ${resource.resourceType.asString} resource in sam for ${googleProject}/${resource.resourceId}"
       )
-      _ <- httpClient.fetch[Unit](
-        Request[F](
-          method = Method.POST,
-          uri = config.samUri
-            .withPath(s"/api/resources/v1/${resource.resourceType.asString}/${resource.resourceId}"),
-          headers = Headers.of(authHeader)
+      _ <- httpClient
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala:126]: `fetch` is deprecated in latest http4s

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala [hunk 6]
```diff
@@ -139,8 +141,7 @@ abstract class BaseCloudServiceRuntimeMonitor[F[_]] {
     runtimeAndRuntimeConfig: RuntimeAndRuntimeConfig,
     dataprocInstances: Set[DataprocInstance], //only applies to dataproc
     message: Option[String],
-    ip: Option[IP] = None,
-    toolsToCheck: List[RuntimeImageType] = List.empty // If empty, then check all tools; if defined, then only check specified tools
+    ip: Option[IP] = None
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala:144]: previously `toolsToCheck` is an optimization that allows us not to re-check a tool if we've already checked it, but now that we're using `streamFUntilDone`, we don't need this anymore, which simplifies the monitoring logic

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala [hunk 7]
```diff
@@ -237,7 +239,7 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
       _ <- if (runtime.asyncRuntimeFields.isDefined) {
         clusterQuery.updateClusterStatus(runtime.id, RuntimeStatus.PreDeleting, ctx.now).transaction >> publisherQueue
           .enqueue1(
-            DeleteRuntimeMessage(runtime.id, Some(ctx.traceId))
+            DeleteRuntimeMessage(runtime.id, req.deleteDisk, Some(ctx.traceId))
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:242]: I thought we were going to detach the disk in front Leo? Or is that a separate PR?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GceInterpreter.scala [hunk 8]
```diff
@@ -303,9 +303,12 @@ class GceInterpreter[F[_]: Parallel: ContextShift: Logger](
...
     params: DeleteRuntimeParams
   )(implicit ev: ApplicativeAsk[F, TraceId]): F[Option[Operation]] =
-    if (params.asyncRuntimeFields.isDefined)
+    if (params.isAsyncRuntimeFields)
       googleComputeService
-        .deleteInstance(params.googleProject, config.gceConfig.zoneName, InstanceName(params.runtimeName.asString))
+        .deleteInstance(params.googleProject,
+                        config.gceConfig.zoneName,
+                        InstanceName(params.runtimeName.asString),
+                        params.autoDeletePersistentDisk.to[Set])
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GceInterpreter.scala:311]: I see, so wb-libs is calling `instanceClient.setDiskAutoDeleteInstance` for each disk then deleting the instance

**Review by rtitle — APPROVED** [review]
LGTM

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala (hunk 7 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:242]: yea, that'll be a separate PR. This is just to support `deleteDisk` flag

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GceInterpreter.scala (hunk 8 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/GceInterpreter.scala:311]: yep...

**Comment by Qi77Qi** [comment]
going to merge this PR as is and bump wb-libs PR in the PR for detaching disk

**Pull request merged**